components/brave_new_tab_ui/hooks/PrefHookManager.ts

```diff
@@ -0,0 +1,104 @@
+     */
+    constructor(getInitialValue: () => Promise<T>, savePref: (key: keyof T, value: T[keyof T]) => void, addListener: (listener: (prefs: T) => void) => void) {
+        this.#savePref = savePref
+
+        addListener((prefs: T) => {
+            const oldPrefs = this.#lastValue ?? {} as Partial<T>
+            this.#lastValue = prefs
+
+            for (const [pref, listeners] of this.#listeners.entries()) {
+                const newValue = prefs[pref]
+                const oldValue = oldPrefs[pref]
+                if (newValue === oldValue) continue
+
+                for (const listener of listeners) { listener(pref, newValue, oldValue) }
+            }
+        })
+        getInitialValue().then(v => this.#lastValue = v)
+    }
+
+    /**
+     * Save a pref.
+     * @param prefName The name of the pref being saved.
+     * @param value The new value for the pref.
+     */
+    savePref<KeyType extends keyof T>(prefName: KeyType, value: T[KeyType]) {
+        this.#savePref(prefName, value)
+    }
+
+    /**
+     * Get the current value of a pref.
+     * @param prefName The name of the pref to get.
+     * @returns The pref value, or undefined, if the PrefHookManager has not received an initial value.
+     */
+    getPref<KeyType extends keyof T>(prefName: KeyType) {
+        return this.#lastValue?.[prefName]
+    }
+
+    /**
+     * Subscribe to changes in a pref value.
+     * @param prefName The name of the pref to listen to.
+     * @param listener A function to be invoked whenever the pref value changes.
+     */
+    addPrefListener<KeyType extends keyof T>(prefName: KeyType, listener: PrefListener<T, KeyType>) {
+        if (!this.#listeners.has(prefName)) { this.#listeners.set(prefName, []) }
+        this.#listeners.get(prefName)!.push(listener)
+    }
+
+    /**
+     * Unsubscribe from changes to a pref value.
+     * @param prefName The name of the pref to remove the listener from.
+     * @param listener The listener to remove.
+     * @returns Whether or not a listener was removed.
+     */
+    removePrefListener<KeyType extends keyof T>(prefName: KeyType, listener: PrefListener<T, KeyType>) {
+        const prefListeners = this.#listeners.get(prefName)
+        if (!prefListeners) return false
+
+        const index = prefListeners.indexOf(listener)
+        prefListeners.splice(index, 1)
+        return index !== -1
+    }
+}
+
+/**
+ * A hook for subscribing and updating prefs. This hook exposes a similar API
+ * to |React.useState|.
+ * @param prefManager The pref manager for this set of prefs.
+ * @param pref The pref to subscribe to.
+ * @returns An array with two items. The first is the current value of the pref,
+ * the second is a function for setting the pref.
+ */
+export const usePref = <OnType, PrefType extends keyof OnType>(prefManager: PrefHookManager<OnType>, pref: PrefType) => {
+    const [value, setValue] = useState(prefManager.getPref(pref))
```

Comment:
Is there a chance that this could be called before the call to `getInitialValue` has come back? I don't see any hook which depends on `getInitialValue`'s Promise resolving. Maybe it's just not possible with the way that we use it right now?

Reply:
There is indeed that chance! But the hook will get notified when `getInitialValue` resolves (via the listener) and update.

Thanks for bringing this up though - it made me realise there's a bug if `getInitialValue` resolves after the hook has been created but before the component is mounted (and `useEffect` runs). Now I update the value at the start of the effect :smile: